pdns/protobuf-dnstap.cc

```diff
@@ -0,0 +1,248 @@
+#include "gettime.hh"
+#include "protobuf-dnstap.hh"
+#include "dnsparser.hh"
+#include "config.h"
+#ifdef PDNS_CONFIG_ARGS
+#include "logger.hh"
+#define WE_ARE_RECURSOR
+#else
+#include "dolog.hh"
+#endif
+
+#define MAXHOSTNAMELEN 256
+
+DnstapProtoBufMessage::DnstapProtoBufMessage(DNSProtoBufMessage::DNSProtoBufMessageType type)
+{
+  setType(type);
+}
+
+void DnstapProtoBufMessage::addRRsFromPacket(const char* packet, const size_t len, bool includeCNAME)
+{
+#ifdef HAVE_PROTOBUF
+  if (len < sizeof(struct dnsheader))
+    return;
+
+  dnstap::Message* message = proto_message.mutable_message();
+  message->set_response_message(packet, len);
+#endif /* HAVE_PROTOBUF */
+}
+
+void DnstapProtoBufMessage::serialize(std::string& data) const
+{
+#ifdef HAVE_PROTOBUF
+  proto_message.SerializeToString(&data);
```

Comment:
You are duplicating of lot of code between `DnstapProtoBufMessage` and `DefaultProtoBufMessage`, how about creating a base class instead of an empty interface?